src/databricks/labs/ucx/source_code/linters/python_ast.py

```diff
@@ -0,0 +1,324 @@
+        yield node
+        for child in node.get_children():
+            yield from cls.walk(child)
+
+
+class MatchingVisitor(TreeVisitor):
+
+    def __init__(self, node_type: type, match_nodes: list[tuple[str, type]]):
+        super()
+        self._matched_nodes: list[NodeNG] = []
+        self._node_type = node_type
+        self._match_nodes = match_nodes
+
+    @property
+    def matched_nodes(self):
+        return self._matched_nodes
+
+    def visit_call(self, node: Call):
+        if self._node_type is not Call:
+            return
+        try:
+            if self._matches(node.func, 0):
+                self._matched_nodes.append(node)
+        except NotImplementedError as e:
+            logger.warning(f"Missing implementation: {e.args[0]}")
+
+    def visit_import(self, node: Import):
+        if self._node_type is not Import:
+            return
+        self._matched_nodes.append(node)
+
+    def visit_importfrom(self, node: ImportFrom):
+        if self._node_type is not ImportFrom:
+            return
+        self._matched_nodes.append(node)
+
+    def _matches(self, node: NodeNG, depth: int):
+        if depth >= len(self._match_nodes):
+            return False
+        name, match_node = self._match_nodes[depth]
+        if not isinstance(node, match_node):
+            return False
+        next_node: NodeNG | None = None
+        if isinstance(node, Attribute):
+            if node.attrname != name:
+                return False
+            next_node = node.expr
+        elif isinstance(node, Name):
+            if node.name != name:
+                return False
+        else:
+            raise NotImplementedError(str(type(node)))
+        if next_node is None:
+            # is this the last node to match ?
+            return len(self._match_nodes) - 1 == depth
+        return self._matches(next_node, depth + 1)
+
+
+class NodeBase(abc.ABC):
+
+    def __init__(self, node: NodeNG):
+        self._node = node
+
+    @property
+    def node(self):
+        return self._node
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: {repr(self._node)}>"
+
+
+class SysPathChange(NodeBase, abc.ABC):
+
+    def __init__(self, node: NodeNG, path: str, is_append: bool):
+        super().__init__(node)
+        self._path = path
+        self._is_append = is_append
+
+    @property
+    def node(self):
+        return self._node
+
+    @property
+    def path(self):
+        return self._path
+
+    @property
+    def is_append(self):
+        return self._is_append
+
+
+# path directly added to sys.path
+class AbsolutePath(SysPathChange):
+    pass
+
+
+# path added to sys.path using os.path.abspath
+class RelativePath(SysPathChange):
+    pass
+
+
+class SysPathVisitor(TreeVisitor):
+
+    def __init__(self):
+        super()
+        self._aliases: dict[str, str] = {}
+        self._sys_path_changes: list[SysPathChange] = []
+
+    @property
+    def sys_path_changes(self):
```

Comment:
Why not simply making the property public?.. otherwise return a copy of a list, as this method is redundant